### DISCUSSION TITLE
How to use multiple instances used in a monorepo (principle of least privilege)

### DISCUSSION THREAD

**Discussion by reegnz** [renovatebot/renovate#24416] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

36.74.0

### If you're self-hosting Renovate, select which platform you are using.

GitHub Enterprise Server

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

We have a monorepo containing infrastructure configuration. In this monorepo we have multiple target environments, and we want to update versions to be deployed into each environment.

Problem is that for some environments (eg. aws govcloud) we are not allowed to use the same credentials to query versions with datasources (eg. aws-machine-image), because of regulatory reasons (principle of least privilege is required, we cannot have a single AWS principal read release versions across all environments).

So  essentially because of principle of least privilege requirements we are not allowed to run a single renovate instance.

Ideally we'd run multiple renovate instances against the same monorepo, one maintaining dependencies in one part of the repo, and another maintaining dependencies in another part of the repo.

As far as I can tell renovate doesn't support this use-case yet, it only supports a single renovate.json.

Any ideas how to tackle this challenge? 
Would renovate team be open to accept a PR where we add support for such behaviour (configure on the server-side a custom location where to look for renovate.json)?

### What you tried so far.

Splitting up the repo, but that goes against our monorepo preference.

**Comment by rarkins** [comment, marked as answer]
I'm fairly uninterested in seeing such capability added natively because it likely will cause complexity both for users as well as for maintainers, while being non applicable to 99.9% of users (or perhaps more).

However you are still welcome to _propose_ specifically how this would work and assess how much source code changes would be required, so we can assess if in fact it's elegant enough to not be a future burden.

You may be able to achieve what you want today with the following:
- Use a different bot identity (either app or user PAT) for each bot
- Do not use repo config (e.g. `renovate.json`) except if it's common for all bots
- Use different global config (e.g. `config.js`) for each bot
- Use a different `branchPrefix` for each bot so that they can ignore each others' branches

**Reply by reegnz** [reply]
So basically instead of repo config just have everything configured in config.js?

I discovered the branchPrefix and also dependencyDashboardTitle to have distinct branches and dashboards per instance.

Anyway, it's already a big help that we could achieve that with some non-conventional config.

My only proposal right now then is to allow a custom `renovate.json` location in `config.js` to override the default repo config name/location. That way we could have multiple repo configs targeting the different renovate instances per environment, without having to move that config into `config.js`.

I'll have to validate, but my intuition tells me that's the only missing part.

**Reply by rarkins** [reply]
The `config.js` is ultimately a JS file which can execute arbitrary code - you might be able to use it to _read_ a file from the repo too. e.g. config can remain in the repo, but just not the conventional way

**Reply by reegnz** [reply]
That's a possible(?) hack/workaround, thanks for pointing that out. Definitely not conventional though and I'd not move away too far from the convention if I can help it. But worst case scenario that helps us somewhat.

We're still looking at alternative options as well (splitting the repo is still on the table as well).

I myself aren't much of a monorepo fan, but it is what it is.
